packages/arb-token-bridge-ui/src/hooks/useSelectedToken.ts

```diff
@@ -0,0 +1,194 @@
+import { useCallback, useMemo } from 'react'
+import useSWRImmutable from 'swr/immutable'
+
+import { useTokenFromSearchParams } from '../components/TransferPanel/TransferPanelUtils'
+import { ERC20BridgeToken, TokenType } from './arbTokenBridge.types'
+import {
+  getL2ERC20Address,
+  isTokenArbitrumOneNativeUSDC,
+  isTokenArbitrumSepoliaNativeUSDC,
+  isTokenMainnetUSDC,
+  isTokenNativeUSDC,
+  isTokenSepoliaUSDC
+} from '../util/TokenUtils'
+import { useNetworks } from './useNetworks'
+import { isNetwork } from '../util/networks'
+import { CommonAddress } from '../util/CommonAddressUtils'
+import { useNetworksRelationship } from './useNetworksRelationship'
+import { Provider } from '@ethersproject/providers'
+import { getChainIdFromProvider } from '@/token-bridge-sdk/utils'
+import {
+  useTokensFromLists,
+  useTokensFromUser
+} from '../components/TransferPanel/TokenSearchUtils'
+
+const commonUSDC = {
+  name: 'USD Coin',
+  type: TokenType.ERC20,
+  symbol: 'USDC',
+  decimals: 6,
+  listIds: new Set<number>()
+}
+
+export const useSelectedToken = () => {
+  const { tokenFromSearchParams, setTokenQueryParam } =
+    useTokenFromSearchParams()
+  const [networks] = useNetworks()
+  const { childChain, childChainProvider, parentChain, parentChainProvider } =
+    useNetworksRelationship(networks)
+  const tokensFromLists = useTokensFromLists()
+  const tokensFromUser = useTokensFromUser()
+
+  const fetcher: () => Promise<ERC20BridgeToken | null> =
+    useCallback(async () => {
+      const tokenAddressLowercased = tokenFromSearchParams?.toLowerCase()
+
+      if (!tokenAddressLowercased) {
+        return null
+      }
+
+      if (isTokenNativeUSDC(tokenAddressLowercased)) {
+        return getUsdcToken({
+          tokenAddress: tokenAddressLowercased,
+          parentChainProvider,
+          childChainProvider
+        })
+      }
+
+      if (!tokensFromLists || !tokensFromUser) {
+        return null
+      }
+
+      return (
+        tokensFromLists[tokenAddressLowercased] ||
+        tokensFromUser[tokenAddressLowercased] ||
+        null
+      )
+    }, [
+      childChainProvider,
+      parentChainProvider,
+      tokenFromSearchParams,
+      tokensFromLists,
+      tokensFromUser
+    ])
+
+  const shouldFetch = useMemo(() => {
+    return (
+      typeof tokensFromLists !== 'undefined' &&
+      typeof tokensFromUser !== 'undefined'
+    )
+  }, [tokensFromLists, tokensFromUser])
```

Comment:
`const shouldFetch = tokensFromLists && tokensFromUsers`

Reply:
Or maybe we can even line it below:
`const queryKey = tokensFromLists && tokensFromUsers ? [parentChain.id, childChain.id, tokenFromSearchParams, 'useSelectedToken'] as const : null`